docs/rfcs/generic-record-sync.md

```diff
@@ -0,0 +1,1287 @@
+          `prefer_deletion` is true
+
+4. The "merged delta" is applied to the mirror record to produce the new mirror
+   record which will be uploaded to the server.
+
+    - This record will have a vector clock that is a descendent of the local,
+      mirror, and incoming remote clocks.
+
+## New metadata records:
+
+These are not stored in meta/blah, they're stored in
+`$collection/__metadata__:blah` (or, `$collection/__metadata__%3Ablah`, really).
+This is to allow transactional behavior during updates, which is important for
+the case of schema changes. Existing clients will need to change to ignore these
+records.
+
+(I'm willing to bikeshed over the name so long as it doesnt fit any pattern any
+current client generates)
+
+**Important**: This requires a change to current clients so that they know to ignore all items
+in a collection whose ID starts with `__metadata__:`!
+
+### `$collection/__metadata__:client_info`
+
+Information about clients. An object with a single field currently, but possibly
+more in the future (the library must make an effort to not drop fields it does not
+understand when updating this record). The only field is `"clients"`, which is
+an array of records, each with the following properties:
+
+- `"id"`: A unique ID generated on DB creation. Unrelated to any sort of current client ID. Discussed in the section on counters/consistency. This is a string.
+
+    - It's illegal for this to be duplicated. (If that happens, the `__metadata__:clients` record is considered corrupted and is discarded, I guess).
+
+- `"native_schema_version"`: This clients "native" schema version for this collection.
+    - This is a semver version string.
+    - This is the latest version it was told to use locally, even if in practice it uses a more up to date schema it fetched. This is effectively the version that the code adding records understands
+
+- `"local_schema_version"`: The latest version of the schema that this client understands.
+    - This is also a semver version string.
+    - This is always semver-compatible with `"native_schema_version"`.
+
+- `"remote_schema_version"`: The last remote schema version it saw.
+    - This is a semver version string.
+    - If this is not semver compatible with `"native_schema_version"`, then the client is locked out and will not sync records.
+        - Note: There are some exceptions, see the section on prerelease version migrations.
+
+- `"metaschema_version"`: The version of the "metaschema", e.g. the schema description format we support. This changes when we add new features to the rust code implementing the schema format.
+    - This is a semver version string, but is unrelated to `{native,local,remote}_schema_version`.
+
+- `"last_sync"`: The most recent X-Weave-Timestamp (as returned by e.g. the fetch to `info/collections` we do before the sync or something). This is for expiring records from this list.
+
+### `$collection/__metadata__:schema`
```

Comment:
Would we expect clients to push new versions of this schema as we deploy new versions of clients?  If so, how do we protect the user's data against corruption by a buggy/malicious client?  I wonder if there's an alternate approach here where we manually publish known-good schema records via some out-of-band mechanism, and just have the clients refer to them by number.  For example we might publish "logins schema v1.2.0" to some well-known location, and then each client can just say "I'm using logins schema X.Y.Z" and count of other clients being able to fetch it, rather than having to push the schema itself up into the remote database.

Reply:
That makes sense and sounds reasonable. I didn't really consider very many server changes, and in general all the ones I considered were non-critical to the general functionality.

The main downside of this is that it's a bit of a hassle to make schema changes, in that it requires contacting us to update the server somehow, but maybe that's fine.

It might eliminate the need for __metadata__:clients as well, or at least simplify it, I have to think more about it.

Reply:
> If so, how do we protect the user's data against corruption by a buggy/malicious client

I don't think we can protect for corruption against malicious clients. Fundamentally, they'll always be able to corrupt any record on the server in any way they want.

For buggy ones, at the moment there's no fix beyond a release that updates the schema. Thankfully, that will work retroactively.

I'm hesitant about a single location for storing the schemas unless it could be made self-serve. If getting started with using this requires cross-team communication I think that makes this a lot less attractive, because part of the motivation is that AFAICT other teams take the path of least resistance here, which tends not to be 'talk to the sync people'. (In the case of containers, they are planning on adding sync, but still never bothered talking to us, even though we explicitly asked them to reach out when they were ready to do this)